packages/base-controller/src/ControllerMessenger.ts

```diff
@@ -210,6 +220,27 @@ export class ControllerMessenger<
     return handler(...params);
   }
 
+  /**
+   * Register a function for getting the initial payload for an event.
+   *
+   * This is used for events that represent a state change, where the payload is the state.
+   * Registering a function for getting the payload allows event selectors to have a point of
+   * comparison the first time state changes.
+   *
+   * @param args - The arguments to this function
+   * @param args.eventType - The event type to register a payload for.
+   * @param args.getPayload - A function for retrieving the event payload.
+   */
+  registerInitialEventPayload<EventType extends Event['type']>({
```

Comment:
It seems that we are only extending the API of the messenger for a very specific use case. Do we expect consumers to call this on its own, or is this really something that we will only call internally? If the former, what are your thoughts on tying this more closely to that use case by adding a fourth argument to `subscribe` that would allow consumers to specify an initial payload getter when subscribing? If the latter, then what do you think about adding a note in the docs here — that if you want to use this for a particular event, make sure to call it ahead of any code that registers subscriptions.

Reply:
We could make it a fourth parameter to `subscribe`, but that'd put the responsibility on each subscriber to know how to get the initial state, and it'd make using selectors a lot more verbose (and would require an additional action to get the initial state).


Whereas with the `registerInitialEventPayload` solution, the publisher of the event is also responsible for registering this "get payload" function. This seems ideal because the publisher is the one who can be most confident what the state will be at any given time, and it keeps the responsibility in a single place.

>Do we expect consumers to call this on its own, or is this really something that we will only call internally?

This is something that would be called in controller constructors, similar to the action registrations. Before any publishing happens.

It doesn't really matter whether this registration occurs before or after subscriptions. It just has to happen before publishing. Just like the action handlers need to be registered before calling.

I'm not sure if it'll have a use beyond the state event. Maybe if we have a controller that is exposing third-party state, it might be relevant (e.g. something that is proxying on-chain or remote API state).